### DISCUSSION TITLE
Deconstruct and using in one line (or without intermediate variable)

### DISCUSSION THREAD

**Discussion by DaveCousineau** [dotnet/csharplang#6001] [Language Ideas]
I completely understand if this is seen as unnecessary, but it would be really nice if I could somehow deconstruct and apply using in the same statement. It could even just be that a using statement can be used as an expression; though I know it's likely that you have reasons for not having done that yet.

Some examples of what I'd like to do:

`using var (x, y) = new SomethingDisposable();`

or

`var (x, y) = using new SomethingDisposable();`

or if the using statement must have the variable part:

`var (x, y) = using var _ = new SomethingDisposable();`

Currently it must look like this:
```
using var sd = new SomethingDisposeable();
var (x, y) = sd;
```

Slightly related, note that it is also not possible to do something like this:
```
var (x, y) = new Something();
using x;
```

`using` statements are probably not so flexible with good reason, so I know this is not likely to be implemented.

**Comment by huoyaoyuan** [comment]
It looks like a feature to `using` a tuple, to `using` two variables.

**Reply by jnm2** [reply]
`using x;` was written in the example, so I don't think both items are IDisposable.

**Reply by huoyaoyuan** [reply]
Yes, I meant "feels like to reader". I'm OK with a feature in that form.

**Reply by DaveCousineau** [reply]
`using x;` is just a side example and not really a part of the suggestion; it's just also something similar that you can't do.

**Reply by jnm2** [reply]
Ah, I think I see what @huoyaoyuan is saying. I was going to say the same thing. I don't like the syntax `using var (x, y) = disposableExpr` because it makes it look like `x` and `y` are both being disposed and that `disposableExpr` is *not* being disposed.

**Comment by Richiban** [comment]
I recently found myself wishing for `using` in patterns, which would nicely extend to your scenario too. (I wanted something likes this:)

```cs 
If (someMethod(out using var disposable))
{
    // disposable is currently in scope and will be disposed at the end of this block
}

```

This could easily be adapted to support your use case too, such as 

```cs
(var x, using var y) = ...
```

**Reply by jnm2** [reply]
The `if` example shows why everyone would trip on this one. `out var` scope does not end at the end of the block mentioned in your comment.

I have liked the idea of `using` in a deconstruction assignment statement.
